File: Sources/AstrolabeProtocol/Core/RuntimeIdentifier.swift
//
//  RuntimeIdentifier.swift
//  astrolabe-protocol
//
//  Created by 轩辕十四 on 2026/7/16.
//

import Foundation

public struct RuntimeOpaqueIdentifier: Codable, Equatable, Hashable, Sendable {
    /// Non-empty wire value whose internal format is owned by its producer.
    public let rawValue: String

    public init(rawValue: String) throws {
        guard !rawValue.isEmpty, rawValue.count <= 256 else {
            throw RuntimeContractValueError.invalidOpaqueIdentifier(rawValue)
        }
        self.rawValue = rawValue
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        try self.init(rawValue: container.decode(String.self))
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        try container.encode(rawValue)
    }
}

public struct RuntimeNamespacedIdentifier: Codable, Equatable, Hashable, Sendable {
    /// Dot-separated identifier that declares its owning namespace.
    public let rawValue: String

    public init(rawValue: String) throws {
        guard Self.isValid(rawValue) else {
            throw RuntimeContractValueError.invalidNamespacedIdentifier(rawValue)
        }
        self.rawValue = rawValue
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        try self.init(rawValue: container.decode(String.self))
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        try container.encode(rawValue)
    }

    public static func isValid(_ value: String) -> Bool {
        guard value.count <= 256 else { return false }
        return value.range(
            of: #"^[a-z][a-z0-9-]*(\.[A-Za-z][A-Za-z0-9_-]*)+$"#,
            options: .regularExpression
        ) != nil
    }
}

public enum RuntimeContractValueError: Error, Equatable, Sendable {
    case invalidOpaqueIdentifier(String)
    case invalidNamespacedIdentifier(String)
    case invalidOpenIdentifier(String)
    case unsafeInteger(Int64)
    case nonFiniteNumber(Double)
}

File: Implementations/Kotlin/src/main/kotlin/dev/astrolabe/protocol/core/RuntimeIdentifier.kt
//
//  RuntimeIdentifier.kt
//  astrolabe-protocol
//
//  Created by 轩辕十四 on 2026/7/20.
//

package dev.astrolabe.protocol

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement

/** Non-empty wire identifier whose internal format is owned by its producer. */
@JvmInline
@Serializable
public value class RuntimeOpaqueIdentifier(
    /** Raw identifier value. */
    public val rawValue: String
) {
    init {
        require(rawValue.isNotEmpty() && rawValue.length <= MAXIMUM_IDENTIFIER_LENGTH) {
            "Opaque identifier must contain between 1 and 256 characters"
        }
    }
}

/** Dot-separated identifier that declares its owning namespace. */
@JvmInline
@Serializable
public value class RuntimeNamespacedIdentifier(
    /** Raw namespaced identifier value. */
    public val rawValue: String
) {
    init {
        require(isValid(rawValue)) { "Namespaced identifier has an invalid format" }
    }

    public companion object {
        /** Returns whether a value satisfies the shared namespaced identifier contract. */
        public fun isValid(value: String): Boolean =
            value.length <= MAXIMUM_IDENTIFIER_LENGTH && NAMESPACED_IDENTIFIER_PATTERN.matches(value)
    }
}

/** Namespaced extension members preserved without platform interpretation. */
@JvmInline
@Serializable
public value class RuntimeExtensionMap(
    /** String-keyed extension values. */
    public val values: Map<String, JsonElement>
) {
    init {
        require(values.keys.all(RuntimeNamespacedIdentifier::isValid)) {
            "Runtime extension keys must be namespaced"
        }
    }
}

private const val MAXIMUM_IDENTIFIER_LENGTH: Int = 256
private val NAMESPACED_IDENTIFIER_PATTERN: Regex =
    Regex("^[a-z][a-z0-9-]*(\\.[A-Za-z][A-Za-z0-9_-]*)+$")
